### PR TITLE
timeFuture should time a closure which errors

### PR DESCRIPTION
**Pull request by Slakah**
For consistency with the other `time*` methods, `timeFuture` should call `stop` on the timing context even if the closure producing the future throws an exception.

### CODE REVIEW

**Comment by erikvanoosten** [comment]
Hi @Slakah. Thanks for your pr!

Method `stop` is already called when the future completes with an error. The API docs for Future.onComplete (http://www.scala-lang.org/api/current/#scala.concurrent.Future) say:

> When this future is completed, either through an exception, or a value, apply the provided function.

However, If you remove your change to `Timer.scala` I will gladly accept the changes to `TimerSpec.scala`.

File: src/test/scala/nl/grons/metrics/scala/TimerSpec.scala [hunk 1]
```diff
@@ -79,6 +79,18 @@ class TimerSpec extends FunSpec with OneInstancePerTest {
       verify(context).stop()
     }
 
+    it("should measure a future closure which errors") {
+      import ExecutionContext.Implicits.global
+      val error = new Exception
+      val caught = intercept[Exception] {
+        timer.timeFuture(throw error)
+      }
+      caught should equal (error)
```
Comment (erikvanoosten) [inline src/test/scala/nl/grons/metrics/scala/TimerSpec.scala]: Can we use `theSameInstanceAs` instead of `equal`?

**Comment by Slakah** [comment]
Hi @erikvanoosten !

> Method stop is already called when the future completes with an error. The API docs for Future.onComplete (http://www.scala-lang.org/api/current/#scala.concurrent.Future) say:
Yeah that piece of functionality works perfectly, for the Future returned by the call-by-name invocation. However it wont stop the timer if the call-by-name function itself throws an exception.

It might be clearer if the test were changed to be more explicit i.e.

```
    it("should measure a future closure which errors") {
      import ExecutionContext.Implicits.global
      val error = new Exception
      val caught = intercept[Exception] {
        timer.timeFuture { throw error }
      }
      caught should equal (error)

      verify(metric).time()
      verify(context).stop()
    }

```

File: src/test/scala/nl/grons/metrics/scala/TimerSpec.scala (hunk 1 above)
Comment (Slakah) [inline src/test/scala/nl/grons/metrics/scala/TimerSpec.scala]: The advantage of checking it's the same object, is that you make sure the original stack trace is not meddled with.

**Comment by erikvanoosten** [comment]
Ah, I get your point now.

I find it hard to decide on whether this is correct or not; I have no use for `timeFuture` in my own projects. @scullxbones @maciej , any insights from your projects?

**Comment by scullxbones** [comment]
I haven't run into this problem, but it's definitely possible.  I think it's a useful patch to fix an edge case.

**Comment by erikvanoosten** [comment]
Alright. @Slakah if you would like to change the test as you suggested I will will merge this. Thanks!

Do you have a pressing need to get this released soon?

**Comment by maciej** [comment]
Makes sense. We never had a case where this would be a problem in our production code, however I can see @Slakah's point.

One comment from my side: should we be catching all `Throwable`s or should we use `NonFatal` extractor?

**Comment by erikvanoosten** [comment]
Thanks @maciej for reminding me. Indeed, @Slakah please use `catch { case NonFatal(e) =>...` 
It probably wont make a lot of difference because we're going to rethrow anyway. But as this whole story is about 'better be safe then sorry' anyway we might as well do it properly all the way.

**Comment by Slakah** [comment]
Thanks for the feedback @maciej and @erikvanoosten, I've implemented the changes as requested.

**Comment by maciej** [comment]
LGTM

**Comment by erikvanoosten** [comment]
Thanks @Slakah.
